test/test-tty-duplicate-key.c

```diff
@@ -0,0 +1,313 @@
+/* Copyright libuv project contributors. All rights reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include "uv.h"
+#include "task.h"
+
+#include <io.h>
+#include <windows.h>
+
+#include <string.h>
+#include <errno.h>
+
+#define ESC "\x1b"
+#define EUR_UTF8 "\xe2\x82\xac"
+#define EUR_UNICODE 0x20AC
+
+
+const char* expect_str = NULL;
+ssize_t expect_nread = 0;
+
+static void dump_str(const char* str, ssize_t len) {
+  ssize_t i;
+  for (i = 0; i < len; i++) {
+    fprintf(stderr, "%#02x ", *(str + i));
+  }
+}
+
+static void print_err_msg(const char* expect, ssize_t expect_len,
+    const char* found, ssize_t found_len) {
+  fprintf(stderr, "expect ");
+  dump_str(expect, expect_len);
+  fprintf(stderr, ", but found ");
+  dump_str(found, found_len);
+  fprintf(stderr, "\n");
+}
+
+static void tty_alloc(uv_handle_t* handle, size_t size, uv_buf_t* buf) {
+  buf->base = malloc(size);
+  buf->len = size;
+}
+
+static void tty_read(uv_stream_t* tty_in, ssize_t nread, const uv_buf_t* buf) {
+  if (nread > 0) {
+    if (nread != expect_nread) {
+      fprintf(stderr, "expected nread %ld, but found %ld\n",
+          (long)expect_nread, (long)nread);
+      print_err_msg(expect_str, expect_nread, buf->base, nread);
+      ASSERT(FALSE);
+    }
+    if (strncmp(buf->base, expect_str, nread) != 0) {
+      print_err_msg(expect_str, expect_nread, buf->base, nread);
+      ASSERT(FALSE);
+    }
+    uv_close((uv_handle_t*) tty_in, NULL);
+  } else {
+    ASSERT(nread == 0);
+  }
+}
+
+static void make_key_event_records(WORD virt_key, DWORD ctr_key_state,
+    BOOL is_wsl, INPUT_RECORD* records) {
```

Comment:
Same comment about lining up the function arguments. I won't keep pointing it out.